services/src/main/java/org/keycloak/services/resources/admin/permissions/UserPermissions.java

```diff
@@ -398,7 +398,26 @@ public boolean canImpersonate() {
             return false;
         }
 
-        return canImpersonate(new DefaultEvaluationContext(identity, session));
+
+        EvaluationContext context;
+
+        ClientModel client = root.adminAuth().getClient();
+
+        if (client!=null) {
+        context = new DefaultEvaluationContext(identity, session) {
+            @Override
+            public Map<String, Collection<String>> getBaseAttributes() {
+                Map<String, Collection<String>> attributes = super.getBaseAttributes();
+                attributes.put("kc.client.id", Arrays.asList(client.getClientId()));
```

Comment:
I think we should have two attributes instead:

* `kc.client.id`: just like it stands today, it would reference the client to which the token was issued for (holder)
* `kc.client.current.id`: the authenticating client as you are proposing

I'm not sure if `kc.client.current.id` is the best name, do you have something better?

The reason for keeping `kc.client.id` as it stands today is to keep the same behavior across the different places we use the policy engine. As well as keep backward compatibility.

The changes I'm proposing would keep the same behavior you are looking for and centralize changes into `DefaultEvaluationContext`.

Reply:
I'm not sure I understand what you're suggesting, in the current state of the code, kc.client.id already mean different things depending on the context the policy engine is being used in. It already corresponds to the calling client ID in the context of other token exchanges:

https://github.com/keycloak/keycloak/blob/160f2dab06ddb8743992c37f1a4afae7a8e6cece/services/src/main/java/org/keycloak/services/resources/admin/permissions/UserPermissions.java#L344

https://github.com/keycloak/keycloak/blob/160f2dab06ddb8743992c37f1a4afae7a8e6cece/services/src/main/java/org/keycloak/services/resources/admin/permissions/IdentityProviderPermissions.java#L187


https://github.com/keycloak/keycloak/blob/160f2dab06ddb8743992c37f1a4afae7a8e6cece/services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java#L345

and to the holder in other context (through DefaultEvaluationContext)


So if I add a new field, say, `kc.exchanging-client.id`, I would have to create a new policy type, say, "Exchanging Client Policy" in the UI, but it means that for some token exchange types (internal-to-internal, naked impersonation), users would have to use the old "Client Policy" to filter on calling client ID, and for non-naked impersonation, they would have to use a new "Exchanging Client Policy".

Reply:
Indeed, it means different things depending on the context the PE is being used. My point is that the client policy could be flexible enough so that you can choose if the validation should happen against the authenticated client or the holder of the token.

By changing the meaning of `kc.client.id` you could be breaking people that are expecting the attribute to reference the holder. That is the only concern I have here, although I think that make impersonation aligned with other forms of token exchange makes a lot of sense.

You would not need to create a new type of policy, but just have a switch in the client policy that tells whether or not the client id should be used based on the holder or the authenticated client.

Makes more sense now? It is mainly about backward compatibility.

Reply:
Alright, I think I understand now, I have to add a new option to "Client Policy" that says something like "Check against token holder"?

What should as use as the default value for this new option? If we make it "true", we break all existing token exchange configurations (because internal-to-internal, naked exhange etc, all check against exchanging client ID), and if we make it "false" we only break people who are currently using a "Client Policy" in the context of a regular, non-naked impersonation, and who explicitely expect the client policy to apply to the client holder, like it currently does. So this latter option seems like the way to go.

Reply:
@maxbes +1. I don't think we are going to break people currently using the client policy as it stands today. By having a default to false we keep current behavior while still allowing people to set the new option if they want to. In this case, the behavior will change as expected ...

Sorry for the late reply.